python/tk_multi_launchapp/software_entity_launcher.py

```diff
@@ -246,55 +232,170 @@ def _build_register_command_data(self, display_name, icon, engine, path, args, v
+        # commands use icons from the engine instead.
+        download_icon_for_commands = []
+
         # List of command data to return
         commands = []
         if path:
+            # A custom application path has been specified in the Software data. If an icon
+            # has also been specified, it will need to be downloaded. Otherwise, attempt to
+            # retrieve the icon from the associated Toolkit engine launcher and use that instead.
+            download_icon = True if icon else False
+            if not icon and engine:
+                sw_versions = self._engine_launcher_software_versions(
+                    engine, display_name, icon, versions
+                )
+                if sw_versions:
+                    self._tk_app.log_debug("Using icon %s from SoftwareVersion for %s." %
+                        ((sw_versions[0].icon), display_name)
+                    )
+                    icon = sw_versions[0].icon
+                else:
+                    self._tk_app.log_debug("No SoftwareVersions found for Toolkit engine %s. "
+                        "Cannot determine icon to display for %s." % (engine, display_name)
+                    )
+
             if versions:
+                # Construct a command for each version.
                 for version in versions:
-                    # Register a command for each version for the path specified.
                     commands.append({
-                        "display_name": display_name, "icon": icon,
-                        "engine": engine, "path": path, "args": args,
-                        "version": version,
+                        "display_name": display_name, "icon": icon, "engine": engine,
+                        "path": path, "args": args, "version": version,
                     })
             else:
-                # Register a launch command for the specified path
+                # Construct a single, version-less command.
+                commands.append({
+                    "display_name": display_name, "icon": icon, "engine": engine,
+                    "path": path, "args": args, "version": None,
+                })
+
+            if download_icon:
+                # The icon field for these commands will need to be updated with the cached file
+                # downloaded from Shotgun.
+                download_icon_for_commands.extend(commands)
+
+        elif engine:
+            # No application path was specified, triggering "auto discovery" mode. Attempt to
+            # find relevant application path(s) from the engine launcher.
+            self._tk_app.log_debug("Using %s engine launcher to find application paths for %s." %
+                (engine, display_name)
+            )
+            sw_versions = self._engine_launcher_software_versions(
+                engine, display_name, icon, versions
+            ) or []
+            for swv in sw_versions:
```

Comment:
nitpick: maybe `for software_version in software_versions` instead - i always find it much harder to read the code when there are variables like `swv`